File: VB/BlazorClientApp/Services/WeatherForecastService.vb
Imports BlazorClientApp.Data
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading
Imports System.Threading.Tasks

Namespace BlazorClientApp.Services
	Public Class WeatherForecastService
		Private temperatureStatistic() As Tuple(Of Integer, Integer) = { Tuple.Create(-15, -1), Tuple.Create(-12, -2), Tuple.Create(-6, 8), Tuple.Create(4, 14), Tuple.Create(11, 21), Tuple.Create(13, 23), Tuple.Create(18, 32), Tuple.Create(16, 26), Tuple.Create(10, 18), Tuple.Create(1, 11), Tuple.Create(-4, 6), Tuple.Create(-13, 1) }

		Private ConditionsForSummary() As Tuple(Of Integer, String) = { Tuple.Create(28, "Hot"), Tuple.Create(14, "Warm"), Tuple.Create(-9, "Cold"), Tuple.Create(-80, "Freezing") }

		Private Shared WeatherTypes() As String = { "Sunny", "Partly cloudy", "Cloudy", "Storm" }

		Private Function CreateForecast() As List(Of WeatherForecast)
			Dim rng = New Random()
			Dim startDate As DateTime = DateTime.Now

			Dim min As Integer = 0
			Dim max As Integer = 0
			Dim temperatureC? As Integer = Nothing

			Return Enumerable.Range(1, 15).Select(Function(index)
				Dim day = startDate.AddDays(index)
				If day.Month - startDate.AddDays(index - 1).Month <> 0 OrElse temperatureC Is Nothing Then
					min = temperatureStatistic(day.Month - 1).Item1
					max = temperatureStatistic(day.Month - 1).Item2
					temperatureC = rng.Next(min, max)
				Else
					temperatureC = temperatureC + rng.Next(-3, 3)
					temperatureC = Math.Max(temperatureC.Value, min)
					temperatureC = Math.Min(temperatureC.Value, max)
				End If
				Dim weatherTypes_Conflict As Integer = If(rng.NextDouble() < 0.5, 0, (If(rng.NextDouble() < 0.5, 1, If(rng.NextDouble() < 0.5, 2, 3))))
				Dim feelTemper As Double = temperatureC.Value - (weatherTypes_Conflict - 1) * 3
				Return New WeatherForecast With {
					.Date = day,
					.TemperatureC = temperatureC.Value,
					.Precipitates = rng.NextDouble() < weatherTypes_Conflict * 0.3,
					.WeatherType = WeatherTypes(weatherTypes_Conflict),
					.Summary = ConditionsForSummary.First(Function(c) c.Item1 <= feelTemper).Item2
				}
			End Function).ToList()
		End Function
		Private Function CreateDetailedForecast() As List(Of WeatherForecast)
			Dim rng = New Random()
			Dim startDate As DateTime = DateTime.Now.Date
			Return Enumerable.Range(1, 15).SelectMany(Function(day)
				Dim dayDate = startDate.AddDays(day)
				Dim avgTemp As Integer = rng.Next(-20, 55)
				Dim minTemp As Integer = Math.Max(-20, avgTemp - 10)
				Dim maxTemp As Integer = Math.Min(55, avgTemp + 10)
				Return Enumerable.Range(0, 24).Select(Function(hour) New WeatherForecast With {
					.Date = dayDate.AddHours(hour),
					.TemperatureC = rng.Next(minTemp, maxTemp),
					.Summary = ConditionsForSummary(rng.Next(ConditionsForSummary.Length)).Item2,
					.WeatherType = WeatherTypes(rng.Next(WeatherTypes.Length)),
					.Precipitates = rng.Next(100) < 30
				})
			End Function).ToList()
		End Function

		Private Property Forecasts() As List(Of WeatherForecast)
		Private Property DetailedForecast() As List(Of WeatherForecast)
		Public Sub New()
			Forecasts = CreateForecast()
			DetailedForecast = CreateDetailedForecast()
		End Sub
		Public Function GetForecastAsync(Optional ByVal ct As CancellationToken = Nothing) As Task(Of IEnumerable(Of WeatherForecast))
			Return Task.FromResult(Forecasts.AsEnumerable())
		End Function
		Public Function GetDetailedForecastAsync(Optional ByVal ct As CancellationToken = Nothing) As Task(Of WeatherForecast())
			Return Task.FromResult(DetailedForecast.ToArray())
		End Function
		Public Function GetSummariesAsync(Optional ByVal ct As CancellationToken = Nothing) As Task(Of String())
			Return Task.FromResult(ConditionsForSummary.Select(Function(v) v.Item2).ToArray())
		End Function
		Public Function GetCloudinessAsync(Optional ByVal ct As CancellationToken = Nothing) As Task(Of IEnumerable(Of String))
			Return Task.FromResult(WeatherTypes.AsEnumerable())
		End Function
		Private Function InsertInternal(ByVal newValue As IDictionary(Of String, Object)) As WeatherForecast()
			Dim dataItem = New WeatherForecast()
			Update(dataItem, newValue)
			Forecasts.Insert(0, dataItem)
			Return Forecasts.ToArray()
		End Function
		Public Function Insert(ByVal newValue As IDictionary(Of String, Object)) As Task(Of WeatherForecast())
			Return Task.FromResult(InsertInternal(newValue))
		End Function
		Private Function RemoveInternal(ByVal dataItem As WeatherForecast) As WeatherForecast()
			Forecasts.Remove(dataItem)
			Return Forecasts.ToArray()
		End Function
		Public Function Remove(ByVal dataItem As WeatherForecast) As Task(Of WeatherForecast())
			Return Task.FromResult(RemoveInternal(dataItem))
		End Function
		Private Function UpdateInternal(ByVal dataItem As WeatherForecast, ByVal newValue As IDictionary(Of String, Object)) As WeatherForecast()
			For Each field In newValue.Keys
				Select Case field
					Case NameOf(dataItem.Date)
						dataItem.Date = DirectCast(newValue(field), DateTime)
					Case NameOf(dataItem.Summary)
						dataItem.Summary = DirectCast(newValue(field), String)
					Case NameOf(dataItem.TemperatureC)
						dataItem.TemperatureC = DirectCast(newValue(field), Integer)
					Case NameOf(dataItem.Precipitates)
						dataItem.Precipitates = DirectCast(newValue(field), Boolean)
					Case NameOf(dataItem.WeatherType)
						dataItem.WeatherType = DirectCast(newValue(field), String)
				End Select
			Next field
			Return Forecasts.ToArray()
		End Function
		Public Function Update(ByVal dataItem As WeatherForecast, ByVal newValue As IDictionary(Of String, Object)) As Task(Of WeatherForecast())
			Return Task.FromResult(UpdateInternal(dataItem, newValue))
		End Function
	End Class

End Namespace

File: CS/BlazorClientApp/Services/WeatherForecastService.cs
using BlazorClientApp.Data;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace BlazorClientApp.Services {
    public class WeatherForecastService {
        Tuple<int, int>[] temperatureStatistic = new Tuple<int, int>[] {
            Tuple.Create(-15, -1),
            Tuple.Create(-12, -2),
            Tuple.Create(-6, 8),
            Tuple.Create(4, 14),
            Tuple.Create(11, 21),
            Tuple.Create(13, 23),
            Tuple.Create(18, 32),
            Tuple.Create(16, 26),
            Tuple.Create(10, 18),
            Tuple.Create(1, 11),
            Tuple.Create(-4, 6),
            Tuple.Create(-13, 1)
        };

        Tuple<int, string>[] ConditionsForSummary = new Tuple<int, string>[] {
            Tuple.Create( 28 , "Hot"),
            Tuple.Create( 14 , "Warm"),
            Tuple.Create( -9 , "Cold"),
            Tuple.Create( -80 , "Freezing")
        };

        private static string[] WeatherTypes = new[] {
            "Sunny", "Partly cloudy", "Cloudy", "Storm"
        };

        private List<WeatherForecast> CreateForecast() {
            var rng = new Random();
            DateTime startDate = DateTime.Now;

            int min = 0;
            int max = 0;
            int? temperatureC = null;

            return Enumerable.Range(1, 15).Select(index => {
                var day = startDate.AddDays(index);
                if (day.Month - startDate.AddDays(index - 1).Month != 0 || temperatureC == null) {
                    min = temperatureStatistic[day.Month - 1].Item1;
                    max = temperatureStatistic[day.Month - 1].Item2;
                    temperatureC = rng.Next(min, max);
                }
                else {
                    temperatureC = temperatureC + rng.Next(-3, 3);
                    temperatureC = Math.Max(temperatureC.Value, min);
                    temperatureC = Math.Min(temperatureC.Value, max);
                }
                int weatherTypes = rng.NextDouble() < 0.5 ? 0 : (rng.NextDouble() < 0.5 ? 1 : rng.NextDouble() < 0.5 ? 2 : 3);
                double feelTemper = temperatureC.Value - (weatherTypes - 1) * 3;
                return new WeatherForecast {
                    Date = day,
                    TemperatureC = temperatureC.Value,
                    Precipitates = rng.NextDouble() < weatherTypes * 0.3,
                    WeatherType = WeatherTypes[weatherTypes],
                    Summary = ConditionsForSummary.First(c => c.Item1 <= feelTemper).Item2
                };
            }).ToList();
        }
        private List<WeatherForecast> CreateDetailedForecast() {
            var rng = new Random();
            DateTime startDate = DateTime.Now.Date;
            return Enumerable.Range(1, 15).SelectMany(day => {
                var dayDate = startDate.AddDays(day);
                int avgTemp = rng.Next(-20, 55);
                int minTemp = Math.Max(-20, avgTemp - 10);
                int maxTemp = Math.Min(55, avgTemp + 10);
                return Enumerable.Range(0, 24).Select(hour => new WeatherForecast {
                    Date = dayDate.AddHours(hour),
                    TemperatureC = rng.Next(minTemp, maxTemp),
                    Summary = ConditionsForSummary[rng.Next(ConditionsForSummary.Length)].Item2,
                    WeatherType = WeatherTypes[rng.Next(WeatherTypes.Length)],
                    Precipitates = rng.Next(100) < 30
                });
            }).ToList();
        }

        private List<WeatherForecast> Forecasts { get; set; }
        private List<WeatherForecast> DetailedForecast { get; set; }
        public WeatherForecastService() {
            Forecasts = CreateForecast();
            DetailedForecast = CreateDetailedForecast();
        }
        public Task<IEnumerable<WeatherForecast>> GetForecastAsync(CancellationToken ct = default) {
            return Task.FromResult(Forecasts.AsEnumerable());
        }
        public Task<WeatherForecast[]> GetDetailedForecastAsync(CancellationToken ct = default) {
            return Task.FromResult(DetailedForecast.ToArray());
        }
        public Task<string[]> GetSummariesAsync(CancellationToken ct = default) {
            return Task.FromResult(ConditionsForSummary.Select(v => v.Item2).ToArray());
        }
        public Task<IEnumerable<string>> GetCloudinessAsync(CancellationToken ct = default) {
            return Task.FromResult(WeatherTypes.AsEnumerable());
        }
        WeatherForecast[] InsertInternal(WeatherForecast newValue) {
            var dataItem = new WeatherForecast();
            Update(dataItem, newValue);
            Forecasts.Insert(0, newValue);
            return Forecasts.ToArray();
        }
        public Task<WeatherForecast[]> Insert(WeatherForecast newValue) {
            return Task.FromResult(InsertInternal(newValue));
        }
        WeatherForecast[] RemoveInternal(WeatherForecast dataItem) {
            Forecasts.Remove(dataItem);
            return Forecasts.ToArray();
        }
        public Task<WeatherForecast[]> Remove(WeatherForecast dataItem) {
            return Task.FromResult(RemoveInternal(dataItem));
        }
        WeatherForecast[] UpdateInternal(WeatherForecast dataItem, WeatherForecast newValue) {
            dataItem.Date = newValue.Date;
            dataItem.Summary = newValue.Summary;
            dataItem.TemperatureC = newValue.TemperatureC;
            dataItem.WeatherType = newValue.WeatherType;
            dataItem.Precipitates = newValue.Precipitates;
            return Forecasts.ToArray();
        }
        public Task<WeatherForecast[]> Update(WeatherForecast dataItem, WeatherForecast newValue) {
            return Task.FromResult(UpdateInternal(dataItem, newValue));
        }
    }

}